### ISSUE TITLE
Potential overflow when merging keys

### ISSUE THREAD

**Issue by adeelhasan** [unlock-protocol/unlock#12553]
I haven't recreated this in the UI, but based on the code for the function _timeMachine, if a key's is merged into a non-expiring key, then line 547 will overflow because it will be adding to uint max. It's worth testing this out in practice, perhaps such a possibility is blocked on the UI.

https://github.com/unlock-protocol/unlock/blob/6abe9661d4c6b3982f9b7af95a48cfa481a9dfdc/smart-contracts/contracts/mixins/MixinKeys.sol#L535-L554

**Comment by adeelhasan** [comment]
there is a gas improvement also in this function, but I wanted to report this possibility. also, I read someplace in the documentation that the smart contracts are updated on quarterly cadence, so no hurry on PR reviews and such. I am spending more time on the MixinKeys file, and will put notes into issues here (or I could send you a google doc.) I will probably also send more UI type notes, and then wait for more feedback on the solidity side. cheers.

**Comment by julien51** [comment]
Thanks for reporting this @adeelhasan ! This is a valid issue! @clemsos will look into this and add a fix. Unlimited/non-expiring keys should not be mergeable, and merging 2 keys which would overflow should fail.

**Comment by adeelhasan** [comment]
Sure, you are welcome.

how non-expiring time can be represented is something to ponder on here are the options I can think of:

1. using max int
- pros: simple to do
- cons: risk of overflow; more gas usage in calldata (evm uses less gas more 0s you use in calldata)

2. using 0
- cons: wont be able to signify actual zero value, and cannot distinguish with uninitialized/default value

3. being explicit about it, ie. have a bool flag "non_expiring"
- pros: able to clearly express the context
- cons: can take up more storage

One solution is to split a storage slot like below. I can also make sense to use uint64 for the value of time, that is far enough in the future for most contexts ... we could use half a slot or something like that.

enum TimeType {
   Always,
   NotAssigned,
   Normal
}

struct Time {
    TimeType type, //will be equivalent to uint8
    uint248 value
}


contract MyContract {
    TimeType expirationDate;
    .....
}
